### PR TITLE
Upgrade Cassandra version to 2.1.10

### PR DESCRIPTION
**Pull request by yuyantingzero**
1. Update Cassandra version from 2.0.16 to 2.1.10
2. Update Cassandra package installation (due to 404 trying to download archive.apache.org).
3. Update parser in cassandra_stress benchmark. The metric name is also changed.
4. Update config files. (cassandra.yaml, cassandra-env.sh)

### CODE REVIEW

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 1]
```diff
@@ -188,36 +191,29 @@ def WaitForLoaderToFinish(vm):
...
+    try:
+      value = regex_util.ExtractGroup(r'%s[\t ]+: ([\d\.:]+)' % metric, resp)
+      if metric == RESULTS_METRICS[-1]:  # Total operation time
+        value = value.split(':')
+        results[metric].append(
+            int(value[0]) * 3600 + int(value[1]) * 60 + int(value[2]))
+      else:
+        results[metric].append(float(value))
+    except regex_util.NoMatchError:
+      logging.exception('No value for %s', metric)
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: When does this occur?

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 2]
```diff
@@ -70,12 +72,18 @@
...
 
 # Stress test options.
-CONSISTENCY_LEVEL = 'quorum'
+CONSISTENCY_LEVEL = 'QUORUM'
 REPLICATION_FACTOR = 3
 RETRIES = 1000
 
 CASSANDRA_STRESS = posixpath.join(cassandra.CASSANDRA_DIR, 'tools', 'bin',
                                   'cassandra-stress')
+RESULTS_METRICS = ['op rate', 'partition rate', 'row rate', 'latency mean',
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Could you comment on what these constants are?
Also, prefer tuples for constants.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 1 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: This doesn't occur in current metrics. 
But the new cassandra stress tool also reports:
"Total partitions", "Total errors", "total gc count", "total gc mb", "total gc time (s)", "avg gc time (ms)", "stdev gc time(ms)". The avg gc time might see "NaN" if no gc occurs during the test. I haven't added these new metrics yet.

**Comment by cmccoy** [comment]
Could you fix the [lint error](https://travis-ci.org/GoogleCloudPlatform/PerfKitBenchmarker/jobs/86733302)?

**Comment by yuyantingzero** [comment]
PTAL, @cmccoy

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 3]
```diff
@@ -70,13 +72,38 @@
...
+                                # the number displayed in the column.
+    'latency 99th percentile',  # 99% of the time the latency was less than
+                                # the number displayed in the column.
+    'latency 99.9th percentile',  # 99.9% of the time the latency was less than
+                                  # the number displayed in the column.
+    'latency max',  # Maximum latency in milliseconds.
+    'Total partitions',  # Number of partitions.
+    'Total errors',  # Number of errors.
+    'Total operation time')  # Total operation time.
+AGGREGATED_METRICS = {'op rate', 'partition rate', 'row rate',
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: A link to the docs is great. Could you briefly document `AGGREGATED_METRICS` and `MAXIMUM_METRICS`?

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 4]
```diff
@@ -118,23 +145,23 @@ def _ResultFilePath(vm):
...
-      '%s '
-      '--file "%s" --nodes %s '
-      '--replication-factor %s --consistency-level %s '
-      '--num-keys %s -K %s -t %s' % (
-          CASSANDRA_STRESS,
-          _ResultFilePath(vm),
-          ','.join(data_node_ips),
-          REPLICATION_FACTOR, CONSISTENCY_LEVEL, FLAGS.num_keys,
-          RETRIES, FLAGS.num_cassandra_stress_threads))
+      '%s write cl=%s n=%s '
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Nit: as this expression becomes more complex, using `str.format` with named arguments would be nice.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 5]
```diff
@@ -164,7 +189,9 @@ def RunCassandraStress(benchmark_spec):
     logging.info('Num keys not set, using %s in cassandra-stress test.',
                  FLAGS.num_keys)
   logging.info('Executing the benchmark.')
-  args = [((loader_vm, data_node_ips), {}) for loader_vm in loader_vms]
+  keys_per_vm = FLAGS.num_keys / len(loader_vms)
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: What if `num_keys` is not evenly divisible by `len(loader_vms)`?

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 1 above)
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Could you separate the metrics which might be missing from the metrics which should always be present?
Also, consider throwing an error if no metrics are found.

Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Removed try, except block.
All current metrics always are present. The missing metrics is "avg gc time" which is not in this PR.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 5 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Added a warning logging here.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 4 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Done.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 3 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Done.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 6]
```diff
@@ -118,23 +148,23 @@ def _ResultFilePath(vm):
...
-      '%s '
-      '--file "%s" --nodes %s '
-      '--replication-factor %s --consistency-level %s '
-      '--num-keys %s -K %s -t %s' % (
-          CASSANDRA_STRESS,
-          _ResultFilePath(vm),
-          ','.join(data_node_ips),
-          REPLICATION_FACTOR, CONSISTENCY_LEVEL, FLAGS.num_keys,
-          RETRIES, FLAGS.num_cassandra_stress_threads))
+      '{0} write cl={1} n={2} '
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: I don't think this is clearer than before. You can pass `format` keyword arguments, like:
```{python}
'-node {node} -schema replication\(factor={replication_factor}\)'.format(
    node=','.join(data_node_ips), replication_factor=REPLICATION_FACTOR)
```
Makes it less likely to yield incorrect results due to an ordering error.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py [hunk 7]
```diff
@@ -118,23 +148,23 @@ def _ResultFilePath(vm):
...
+      '-node {3} -schema replication\(factor={4}\) -pop seq={5}..{6} '
+      '-log file={7} -rate threads={8} -errors retries={9}'.format(
+          CASSANDRA_STRESS, CONSISTENCY_LEVEL, keys_per_vm,
+          ','.join(data_node_ips), REPLICATION_FACTOR,
+          loader_index * keys_per_vm + 1, (loader_index + 1) * keys_per_vm,
+          _ResultFilePath(vm), RETRIES, FLAGS.num_cassandra_stress_threads))
```
Comment (cmccoy) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: It looks like threads and retries are swapped here. See above :)

Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Done.

File: perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py (hunk 6 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py]: Done.

**Comment by cmccoy** [comment]
LGTM. Thanks.

File: CHANGES.md [hunk 8]
```diff
@@ -1,3 +1,8 @@
+# v0.24.0
+
+* Enhancements:
+  * **Breaking Change** Update cassandra version to 2.1.10 (GH-577)
```
Comment (cmccoy) [inline CHANGES.md]: Could you describe how this is breaking?

Comment (yuyantingzero) [inline CHANGES.md]: I will update CHANGES.md in next release. Basically, the cassandra stress tool is upgraded with the cassandra package. And it is reporting a set of new metrics now and some of the old metrics are also renamed.

**Pull request merged**